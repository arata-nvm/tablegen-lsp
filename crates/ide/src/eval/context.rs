use ecow::{eco_format, EcoString};
use syntax::parser::TextRange;

use crate::{
    file_system::{FileId, FileRange},
    handlers::diagnostics::Diagnostic,
    symbol_map::{symbol::SymbolId, variable::VariableId, SymbolMap},
};

use super::{scope::Scopes, EvalDatabase, Evaluation};

pub struct EvalCtx<'a> {
    pub db: &'a dyn EvalDatabase,
    pub file_trace: Vec<FileId>,
    pub symbol_map: SymbolMap,
    pub diagnostics: Vec<Diagnostic>,
    pub scopes: Scopes,
    pub anonymous_def_index: u32,
}

impl<'a> EvalCtx<'a> {
    pub fn new(db: &'a dyn EvalDatabase, root_file: FileId) -> Self {
        Self {
            db,
            file_trace: vec![root_file],
            symbol_map: SymbolMap::default(),
            diagnostics: Vec::new(),
            scopes: Scopes::default(),
            anonymous_def_index: 0,
        }
    }

    pub fn current_file_id(&self) -> FileId {
        *self.file_trace.last().expect("file_trace is empty")
    }

    pub fn push_file(&mut self, file_id: FileId) {
        self.file_trace.push(file_id);
    }

    pub fn pop_file(&mut self) {
        self.file_trace.pop().expect("file_trace is empty");
    }

    pub fn resolve_id(&self, name: &EcoString) -> Option<SymbolId> {
        if let Some(def_id) = self.symbol_map.find_def(name) {
            return Some(def_id.into());
        }
        if let Some(symbol_id) = self.scopes.find_local(name) {
            return Some(symbol_id);
        }
        None
    }

    pub fn resolve_id_in_current_scope(&self, name: &EcoString) -> Option<VariableId> {
        self.scopes.find_variable_in_current_scope(name)
    }

    pub fn error(&mut self, range: TextRange, message: impl Into<String>) {
        let file = self.current_file_id();
        self.diagnostics
            .push(Diagnostic::new(FileRange::new(file, range), message));
    }

    pub fn finish(self) -> Evaluation {
        Evaluation {
            symbol_map: self.symbol_map,
            diagnostics: self.diagnostics,
        }
    }

    pub fn next_anonymous_def_name(&mut self) -> EcoString {
        let index = self.anonymous_def_index;
        self.anonymous_def_index += 1;
        eco_format!("anonymous_{index}")
    }
}
