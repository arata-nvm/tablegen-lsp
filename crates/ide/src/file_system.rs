use std::collections::{HashMap, VecDeque};
use std::path::{Path, PathBuf};

use ecow::EcoString;
use syntax::ast::AstNode;
use syntax::parser::{TextRange, TextSize};
use syntax::{SyntaxNode, SyntaxNodePtr, ast};

use crate::db::{Db, parse};

#[derive(Debug, Clone, Copy, Ord, PartialOrd, Eq, PartialEq, Hash)]
pub struct FileId(pub u32);

#[derive(Debug, Clone, Copy, Eq, PartialEq, Hash)]
pub struct FilePosition {
    pub file: FileId,
    pub position: TextSize,
}

impl FilePosition {
    pub fn new(file: FileId, position: TextSize) -> Self {
        Self { file, position }
    }
}

#[derive(Debug, Clone, Copy, Eq, PartialEq, Hash)]
pub struct FileRange {
    pub file: FileId,
    pub range: TextRange,
}

impl FileRange {
    pub fn new(file: FileId, range: TextRange) -> Self {
        Self { file, range }
    }
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct FilePath(pub PathBuf);

impl FilePath {
    pub fn join(&self, path: impl AsRef<Path>) -> FilePath {
        FilePath(self.0.join(path))
    }

    pub fn parent(&self) -> Option<FilePath> {
        self.0.parent().map(FilePath::from)
    }
}

impl From<&Path> for FilePath {
    fn from(value: &Path) -> Self {
        Self(value.to_path_buf())
    }
}

#[derive(Debug, Default)]
pub struct FileSet {
    path_to_id: HashMap<FilePath, FileId>,
    id_to_path: HashMap<FileId, FilePath>,
}

impl FileSet {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn insert(&mut self, file_id: FileId, path: FilePath) {
        self.path_to_id.insert(path.clone(), file_id);
        self.id_to_path.insert(file_id, path.clone());
    }

    pub fn remove(&mut self, file_id: &FileId) {
        if let Some(path) = self.id_to_path.remove(file_id) {
            self.path_to_id.remove(&path);
        }
    }

    pub fn contains(&self, file_id: &FileId) -> bool {
        self.id_to_path.contains_key(file_id)
    }

    pub fn file_for_path(&self, path: &FilePath) -> Option<FileId> {
        self.path_to_id.get(path).copied()
    }

    pub fn path_for_file(&self, file_id: &FileId) -> &FilePath {
        &self.id_to_path[file_id]
    }

    pub fn iter_files(&self) -> impl Iterator<Item = FileId> + '_ {
        self.id_to_path.keys().copied()
    }
}

#[derive(Debug, Clone, Copy, Ord, PartialOrd, Eq, PartialEq, Hash)]
pub struct SourceUnitId(FileId);

impl SourceUnitId {
    pub fn from_root_file(root_file: FileId) -> Self {
        Self(root_file)
    }
}

#[derive(Debug)]
pub struct SourceUnit {
    root: FileId,
    includes: HashMap<FileId, IncludeMap>,
}

impl SourceUnit {
    pub fn root(&self) -> FileId {
        self.root
    }

    pub fn iter_files(&self) -> impl Iterator<Item = FileId> + '_ {
        self.includes.keys().copied()
    }

    pub fn include_map(&self, file_id: &FileId) -> Option<&IncludeMap> {
        self.includes.get(file_id)
    }
}

pub trait FileSystem {
    fn assign_or_get_file_id(&mut self, path: FilePath) -> FileId;

    fn path_for_file(&self, file_id: &FileId) -> &FilePath;

    fn read_content(&self, file_path: &FilePath) -> Option<String>;
}

pub fn collect_sources<FS: FileSystem>(
    db: &mut dyn Db,
    fs: &mut FS,
    root: FileId,
    include_dirs: &[FilePath],
) -> SourceUnit {
    let mut includes = HashMap::new();

    let mut queue = VecDeque::new();
    queue.push_back(root);
    while let Some(file_id) = queue.pop_front() {
        if includes.contains_key(&file_id) {
            continue;
        }

        let parse = parse(db, file_id);

        let file_path = fs.path_for_file(&file_id);
        let file_dir = file_path.parent().expect("file dir not found");

        let mut include_dir_list = vec![file_dir];
        include_dir_list.extend(include_dirs.iter().cloned());

        let mut include_map = HashMap::new();
        for (include_id, include_path) in list_includes(parse.syntax_node()) {
            let Some(resolved_file_id) =
                resolve_include_file(db, fs, include_path, &include_dir_list)
            else {
                continue;
            };

            include_map.insert(include_id, resolved_file_id);
            queue.push_back(resolved_file_id);
        }
        includes.insert(file_id, include_map);
    }

    SourceUnit { root, includes }
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct IncludeId(pub SyntaxNodePtr);

pub type IncludeMap = HashMap<IncludeId, FileId>;

fn list_includes(root_node: SyntaxNode) -> Vec<(IncludeId, EcoString)> {
    (|| -> Option<_> {
        let source_file = ast::SourceFile::cast(root_node)?;
        let stmt_list = source_file.statement_list()?;
        let include_paths = stmt_list
            .statements()
            .filter_map(|stmt| match stmt {
                ast::Statement::Include(include) => {
                    let id = IncludeId(SyntaxNodePtr::new(include.syntax()));
                    let path = include.path()?.value();
                    Some((id, path))
                }
                _ => None,
            })
            .collect();
        Some(include_paths)
    })()
    .unwrap_or_default()
}

fn resolve_include_file<FS: FileSystem>(
    db: &mut dyn Db,
    fs: &mut FS,
    include_path: EcoString,
    include_dir_list: &[FilePath],
) -> Option<FileId> {
    for include_dir in include_dir_list {
        let candidate_file_path = include_dir.join(include_path.as_str());
        if let Some(file_content) = fs.read_content(&candidate_file_path) {
            let file_id = fs.assign_or_get_file_id(candidate_file_path.clone());
            db.set_file_content(file_id, file_content.as_str());
            return Some(file_id);
        }
    }
    None
}
