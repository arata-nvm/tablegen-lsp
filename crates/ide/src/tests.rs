use std::{collections::HashMap, path::Path, sync::Arc};

use crate::{
    db::{SourceDatabase, SourceDatabaseStorage},
    eval::EvalDatabaseStorage,
    file_system::{self, FileId, FilePath, FileSet, FileSystem},
};

const DEFAULT_FILE_PATH: &str = "/main.td";

pub fn single_file(fixture: &str) -> (TestDatabase, Fixture) {
    let mut f = Fixture::single_file(fixture);
    let db = TestDatabase::new(&mut f);
    (db, f)
}

pub fn multiple_files(fixture: &str) -> (TestDatabase, Fixture) {
    let mut f = Fixture::multiple_files(fixture);
    let db = TestDatabase::new(&mut f);
    (db, f)
}

#[salsa::database(SourceDatabaseStorage, EvalDatabaseStorage)]
#[derive(Default)]
pub struct TestDatabase {
    storage: salsa::Storage<Self>,
}

impl salsa::Database for TestDatabase {}

impl TestDatabase {
    fn new(f: &mut Fixture) -> Self {
        let mut db = Self::default();
        for (file_id, content) in f.files() {
            db.set_file_content(file_id, Arc::from(content));
        }

        let source_root = file_system::collect_sources(&mut db, f, f.root_file());
        db.set_source_root(Arc::new(source_root));

        db
    }
}

#[derive(Debug, Default)]
pub struct Fixture {
    file_contents: HashMap<FilePath, String>,
    file_ids: Vec<FileId>,

    file_set: FileSet,
    next_file_id: u32,
}

impl Fixture {
    fn single_file(fixture: &str) -> Self {
        let mut this = Self::default();
        this.insert_file(
            FilePath::from(Path::new(DEFAULT_FILE_PATH)),
            fixture.to_string(),
        );
        this
    }

    fn multiple_files(fixture: &str) -> Self {
        let mut this = Self::default();

        let mut lines = fixture
            .lines()
            .skip_while(|line| !line.starts_with("; "))
            .peekable();
        while let Some(line) = lines.next() {
            let path_str = line
                .strip_prefix("; ")
                .expect("expected header line to start with '; '");
            let path = FilePath::from(Path::new(path_str));

            let mut content = String::new();
            while let Some(line) = lines.peek() {
                if line.starts_with("; ") {
                    break;
                }

                let line = lines.next().expect("line must be present");
                content.push_str(line);
                content.push_str("\n");
            }

            this.insert_file(path, content);
        }

        this
    }

    fn files(&self) -> Vec<(FileId, &str)> {
        self.file_contents
            .iter()
            .map(|(path, content)| {
                (
                    self.file_set
                        .file_for_path(path)
                        .expect("file id must exist"),
                    content.as_str(),
                )
            })
            .collect()
    }

    fn root_file(&self) -> FileId {
        self.file_ids
            .first()
            .expect("at least one file must be exist")
            .clone()
    }

    fn insert_file(&mut self, path: FilePath, content: String) {
        let res = self.file_contents.insert(path.clone(), content);
        assert!(res.is_none(), "duplicate file path");

        let file_id = self.alloc_file_id();
        self.file_set.insert(file_id, path);
        self.file_ids.push(file_id);
    }

    fn alloc_file_id(&mut self) -> FileId {
        let file_id = FileId(self.next_file_id);
        self.next_file_id += 1;
        file_id
    }
}

impl FileSystem for Fixture {
    fn assign_or_get_file_id(&mut self, path: FilePath) -> FileId {
        match self.file_set.file_for_path(&path) {
            Some(file_id) => file_id,
            None => {
                let file_id = self.alloc_file_id();
                self.file_set.insert(file_id, path);
                file_id
            }
        }
    }

    fn path_for_file(&self, file_id: &FileId) -> Option<FilePath> {
        self.file_set.path_for_file(file_id)
    }

    fn read_content(&self, file_path: &FilePath) -> Option<String> {
        self.file_contents.get(file_path).cloned()
    }
}
