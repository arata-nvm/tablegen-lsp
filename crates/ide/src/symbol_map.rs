use std::fmt::Debug;
use std::ops::DerefMut;
use std::{collections::HashMap, ops::Deref};

use ecow::EcoString;
use id_arena::{Arena, Id};
use syntax::parser::TextSize;

use crate::file_system::{FileId, FilePosition, FileRange};

#[derive(Debug, Eq, PartialEq)]
pub struct Class {
    pub name: EcoString,
    pub define_loc: FileRange,
    pub reference_locs: Vec<FileRange>,
    pub template_arg_list: Vec<TemplateArgumentId>,
    pub parent_class_list: Vec<ClassId>,
}

pub type ClassId = Id<Class>;

impl Class {
    pub fn new(
        name: EcoString,
        define_loc: FileRange,
        template_arg_list: Vec<TemplateArgumentId>,
        parent_class_list: Vec<ClassId>,
    ) -> Self {
        Self {
            name,
            define_loc,
            reference_locs: Vec::new(),
            template_arg_list,
            parent_class_list,
        }
    }
}

#[derive(Debug, Eq, PartialEq)]
pub struct TemplateArgument {
    pub name: EcoString,
    pub define_loc: FileRange,
}

pub type TemplateArgumentId = Id<TemplateArgument>;

impl TemplateArgument {
    pub fn new(name: EcoString, define_loc: FileRange) -> Self {
        Self { name, define_loc }
    }
}

#[derive(Debug, Eq, PartialEq)]
pub enum Symbol<'a> {
    Class(&'a Class),
    TemplateArgument(&'a TemplateArgument),
}

impl<'a> Symbol<'a> {
    pub fn name(&self) -> &EcoString {
        match self {
            Symbol::Class(class) => &class.name,
            Symbol::TemplateArgument(template_arg) => &template_arg.name,
        }
    }

    pub fn define_loc(&self) -> &FileRange {
        match self {
            Symbol::Class(class) => &class.define_loc,
            Symbol::TemplateArgument(template_arg) => &template_arg.define_loc,
        }
    }

    pub fn as_class(&self) -> Option<&Class> {
        match self {
            Symbol::Class(class) => Some(class),
            _ => None,
        }
    }

    pub fn as_template_argument(&self) -> Option<&TemplateArgument> {
        match self {
            Symbol::TemplateArgument(template_arg) => Some(template_arg),
            _ => None,
        }
    }
}

#[derive(Debug, Eq, PartialEq, Clone, Copy, Hash, PartialOrd, Ord)]
pub enum SymbolId {
    ClassId(ClassId),
    TemplateArgumentId(TemplateArgumentId),
}

#[derive(Debug, Default, Eq, PartialEq)]
pub struct SymbolMap {
    class_list: Arena<Class>,
    template_arg_list: Arena<TemplateArgument>,
    file_to_class_list: HashMap<FileId, Vec<ClassId>>,
    pos_to_symbol_map: HashMap<FileId, IntervalMap<TextSize, SymbolId>>,
}

// immutable api
impl SymbolMap {
    pub fn new() -> Self {
        Self::default()
    }
}

// mutable api
impl SymbolMap {
    pub fn add_class(&mut self, class: Class, file_id: FileId) -> ClassId {
        let define_loc = class.define_loc;
        let id = self.class_list.alloc(class);
        self.file_to_class_list.entry(file_id).or_default().push(id);
        self.pos_to_symbol_map
            .entry(file_id)
            .or_insert_with(IntervalMap::new)
            .insert(define_loc.range.into(), SymbolId::ClassId(id));
        id
    }

    pub fn add_class_reference(&mut self, class_id: ClassId, reference_loc: FileRange) {
        let class = self.class_mut(class_id);
        class.reference_locs.push(reference_loc);
        self.pos_to_symbol_map
            .entry(reference_loc.file)
            .or_insert_with(IntervalMap::new)
            .insert(reference_loc.range.into(), SymbolId::ClassId(class_id));
    }

    pub fn add_template_argument(
        &mut self,
        template_arg: TemplateArgument,
        file_id: FileId,
    ) -> TemplateArgumentId {
        let define_loc = template_arg.define_loc;
        let id = self.template_arg_list.alloc(template_arg);
        self.pos_to_symbol_map
            .entry(file_id)
            .or_insert_with(IntervalMap::new)
            .insert(define_loc.range.into(), SymbolId::TemplateArgumentId(id));
        id
    }

    pub fn class(&self, class_id: ClassId) -> &Class {
        self.class_list.get(class_id).expect("invalid class id")
    }

    pub fn class_mut(&mut self, class_id: ClassId) -> &mut Class {
        self.class_list.get_mut(class_id).expect("invalid class id")
    }

    pub fn iter_class(&self) -> impl Iterator<Item = (ClassId, &Class)> {
        self.class_list.iter()
    }

    pub fn iter_class_in(&self, file_id: FileId) -> Option<impl Iterator<Item = ClassId>> {
        self.file_to_class_list
            .get(&file_id)
            .cloned()
            .map(|class_list| class_list.into_iter())
    }

    pub fn find_symbol_at(&self, pos: FilePosition) -> Option<Symbol> {
        let id = self
            .pos_to_symbol_map
            .get(&pos.file)
            .and_then(|map| map.values_overlap(pos.position).next().cloned())?;

        match id {
            SymbolId::ClassId(class_id) => Some(Symbol::Class(self.class(class_id))),
            SymbolId::TemplateArgumentId(template_arg_id) => {
                Some(Symbol::TemplateArgument(self.template_arg(template_arg_id)))
            }
        }
    }

    pub fn template_arg(&self, template_arg_id: TemplateArgumentId) -> &TemplateArgument {
        self.template_arg_list
            .get(template_arg_id)
            .expect("invalid template argument id")
    }
}

// FIXME
// because iset::IntervalMap does not implement PartialEq and Eq, we have to implement them
#[derive(Debug, Default)]
struct IntervalMap<T: PartialOrd + Copy + Debug, V: Debug + PartialEq + Eq>(
    iset::IntervalMap<T, V>,
);

impl<T: PartialOrd + Copy + Debug, V: Debug + PartialEq + Eq> IntervalMap<T, V> {
    fn new() -> Self {
        Self(iset::IntervalMap::new())
    }
}

impl<T: PartialOrd + Copy + Debug, V: Debug + PartialEq + Eq> Deref for IntervalMap<T, V> {
    type Target = iset::IntervalMap<T, V>;
    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl<T: PartialOrd + Copy + Debug, V: Debug + PartialEq + Eq> DerefMut for IntervalMap<T, V> {
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.0
    }
}

impl<T: PartialOrd + Copy + Debug, V: Debug + PartialEq + Eq> PartialEq for IntervalMap<T, V> {
    fn eq(&self, other: &Self) -> bool {
        if self.0.len() != other.0.len() {
            return false;
        }

        for ((range1, value1), (range2, value2)) in self.0.iter(..).zip(other.0.iter(..)) {
            if range1 != range2 || value1 != value2 {
                return false;
            }
        }
        true
    }
}

impl<T: PartialOrd + Copy + Debug, V: Debug + PartialEq + Eq> Eq for IntervalMap<T, V> {}
