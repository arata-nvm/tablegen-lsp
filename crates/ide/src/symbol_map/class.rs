use ecow::EcoString;
use id_arena::Id;
use indexmap::IndexMap;

use crate::file_system::FileRange;

use super::{record::RecordFieldId, template_arg::TemplateArgumentId, SymbolMap};

pub type ClassId = Id<Class>;

#[derive(Debug, Clone, Eq, PartialEq)]
pub struct Class {
    pub name: EcoString,
    pub name_to_template_arg: IndexMap<EcoString, TemplateArgumentId>,
    pub name_to_record_field: IndexMap<EcoString, RecordFieldId>,
    pub parent_list: Vec<ClassId>,

    pub define_loc: FileRange,
    pub reference_locs: Vec<FileRange>,
}

impl Class {
    pub fn new(name: EcoString, define_loc: FileRange) -> Self {
        Self {
            name,
            name_to_template_arg: IndexMap::new(),
            name_to_record_field: IndexMap::new(),
            parent_list: Vec::new(),
            define_loc,
            reference_locs: Vec::new(),
        }
    }

    pub fn add_template_arg(&mut self, name: EcoString, template_arg_id: TemplateArgumentId) {
        self.name_to_template_arg.insert(name, template_arg_id);
    }

    pub fn iter_template_arg(&self) -> impl Iterator<Item = TemplateArgumentId> + '_ {
        self.name_to_template_arg.values().copied()
    }

    pub fn find_template_arg(&self, name: &EcoString) -> Option<TemplateArgumentId> {
        self.name_to_template_arg.get(name).copied()
    }

    pub fn add_record_field(&mut self, name: EcoString, record_field_id: RecordFieldId) {
        self.name_to_record_field.insert(name, record_field_id);
    }

    pub fn iter_field(&self) -> impl Iterator<Item = RecordFieldId> + '_ {
        self.name_to_record_field.values().copied()
    }

    pub fn find_field(&self, symbol_map: &SymbolMap, name: &EcoString) -> Option<RecordFieldId> {
        if let Some(field_id) = self.name_to_record_field.get(name) {
            return Some(*field_id);
        }

        for parent_id in &self.parent_list {
            let parent = symbol_map.class(*parent_id);
            if let Some(field_id) = parent.find_field(symbol_map, name) {
                return Some(field_id);
            }
        }

        None
    }

    pub fn add_parent(&mut self, parent_id: ClassId) {
        self.parent_list.push(parent_id);
    }

    pub fn is_subclass_of(&self, symbol_map: &SymbolMap, other_id: ClassId) -> bool {
        if self.parent_list.contains(&other_id) {
            return true;
        }

        for parent_id in &self.parent_list {
            let parent = symbol_map.class(*parent_id);
            if parent.is_subclass_of(symbol_map, other_id) {
                return true;
            }
        }

        false
    }
}
