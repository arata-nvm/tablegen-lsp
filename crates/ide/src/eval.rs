use std::sync::Arc;

use ecow::EcoString;

use syntax::ast;
use syntax::ast::AstNode;
use syntax::ast::SourceFile;

use crate::db::SourceDatabase;
use crate::file::FileId;
use crate::symbol_map::{Class, SymbolMap};

#[salsa::query_group(EvalDatabaseStorage)]
pub trait EvalDatabase: SourceDatabase {
    fn symbol_map(&self, file_id: FileId) -> Arc<SymbolMap>;
}

fn symbol_map(db: &dyn EvalDatabase, file_id: FileId) -> Arc<SymbolMap> {
    let parse = db.parse(file_id);
    let source_file = ast::SourceFile::cast(parse.syntax_node()).unwrap();
    Arc::new(eval(source_file))
}

fn eval(source_file: SourceFile) -> SymbolMap {
    let mut ctx = EvalCtx::new();
    source_file.eval(&mut ctx);
    ctx.symbol_map
}

#[derive(Debug, Default)]
pub struct EvalCtx {
    symbol_map: SymbolMap,
}

impl EvalCtx {
    pub fn new() -> Self {
        Self::default()
    }
}

pub trait Eval {
    type Output;
    fn eval(self, ctx: &mut EvalCtx) -> Option<Self::Output>;
}

impl Eval for ast::SourceFile {
    type Output = ();
    fn eval(self, ctx: &mut EvalCtx) -> Option<Self::Output> {
        self.statement_list()?.eval(ctx);
        Some(())
    }
}

impl Eval for ast::StatementList {
    type Output = ();
    fn eval(self, ctx: &mut EvalCtx) -> Option<Self::Output> {
        for stmt in self.statements() {
            stmt.eval(ctx);
        }
        Some(())
    }
}

impl Eval for ast::Statement {
    type Output = ();
    fn eval(self, ctx: &mut EvalCtx) -> Option<Self::Output> {
        match self {
            ast::Statement::Class(class) => class.eval(ctx),
            _ => unimplemented!(),
        }
    }
}

impl Eval for ast::Class {
    type Output = ();
    fn eval(self, ctx: &mut EvalCtx) -> Option<Self::Output> {
        let name = self.name()?.eval(ctx)?;
        let class = Class::new(name);
        ctx.symbol_map.add_class(class);
        Some(())
    }
}

impl Eval for ast::Identifier {
    type Output = EcoString;
    fn eval(self, _ctx: &mut EvalCtx) -> Option<Self::Output> {
        self.value()
    }
}
