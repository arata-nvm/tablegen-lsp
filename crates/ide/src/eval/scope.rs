use std::collections::HashMap;

use ecow::EcoString;

use crate::symbol_map::{
    class::ClassId,
    def::DefId,
    record::Record,
    symbol::SymbolId,
    variable::{Variable, VariableId},
    SymbolMap,
};

#[derive(Debug)]
pub struct Scopes {
    scopes: Vec<Scope>,
}

impl Default for Scopes {
    fn default() -> Self {
        Self {
            scopes: vec![Scope::new(ScopeKind::Root)],
        }
    }
}

impl Scopes {
    pub fn push(&mut self, kind: ScopeKind) {
        self.scopes.push(Scope::new(kind));
    }

    pub fn pop(&mut self) -> Scope {
        self.scopes.pop().expect("scope is empty")
    }

    pub fn current_record_id(&self) -> SymbolId {
        self.scopes
            .iter()
            .rev()
            .find_map(|scope| scope.record_id())
            .expect("scope is empty")
    }

    pub fn current_record(&self) -> &Record {
        self.scopes
            .iter()
            .rev()
            .find_map(|scope| scope.record())
            .expect("scope is empty")
    }

    pub fn current_record_mut(&mut self) -> &mut Record {
        self.scopes
            .iter_mut()
            .rev()
            .find_map(|record| record.record_mut())
            .expect("scope is empty")
    }

    pub fn add_variable(&mut self, symbol_map: &mut SymbolMap, variable: Variable) {
        let name = variable.name.clone();
        let id = symbol_map.add_variable(variable);
        let current_scope = self.scopes.last_mut().expect("scope is empty");
        current_scope.name_to_variable.insert(name, id);
    }

    pub fn find_local(&self, name: &EcoString) -> Option<SymbolId> {
        for scope in self.scopes.iter().rev() {
            if let Some(id) = scope.find_variable(name) {
                return Some(id.into());
            }
            if let Some(record) = scope.record() {
                if let Some(field_id) = record.find_field(name) {
                    return Some(field_id.into());
                }
                if let Some(template_arg_id) = record.find_template_arg(name) {
                    return Some(template_arg_id.into());
                }
            }
        }
        None
    }

    pub fn find_variable_in_current_scope(&self, name: &EcoString) -> Option<VariableId> {
        let scope = self.scopes.last().expect("scope is empty");
        scope.find_variable(name)
    }
}

#[derive(Debug)]
pub struct Scope {
    pub kind: ScopeKind,
    pub name_to_variable: HashMap<EcoString, VariableId>,
}

#[derive(Debug)]
pub enum ScopeKind {
    Root,
    Class(ClassId, Record),
    Def(DefId, Record),
    Foreach(EcoString, VariableId),
}

impl Scope {
    pub fn new(kind: ScopeKind) -> Self {
        Self {
            kind,
            name_to_variable: HashMap::new(),
        }
    }

    pub fn record_id(&self) -> Option<SymbolId> {
        match self.kind {
            ScopeKind::Root => None,
            ScopeKind::Class(id, _) => Some(id.into()),
            ScopeKind::Def(id, _) => Some(id.into()),
            ScopeKind::Foreach(_, _) => None,
        }
    }

    pub fn record(&self) -> Option<&Record> {
        match &self.kind {
            ScopeKind::Root => None,
            ScopeKind::Class(_, record) => Some(record),
            ScopeKind::Def(_, record) => Some(record),
            ScopeKind::Foreach(_, _) => None,
        }
    }

    pub fn record_mut(&mut self) -> Option<&mut Record> {
        match &mut self.kind {
            ScopeKind::Root => None,
            ScopeKind::Class(_, record) => Some(record),
            ScopeKind::Def(_, record) => Some(record),
            ScopeKind::Foreach(_, _) => None,
        }
    }

    pub fn into_class(self) -> (ClassId, Record) {
        match self.kind {
            ScopeKind::Class(id, class) => (id, class),
            _ => panic!("not a class"),
        }
    }

    pub fn into_def(self) -> (DefId, Record) {
        match self.kind {
            ScopeKind::Def(id, def) => (id, def),
            _ => panic!("not a def"),
        }
    }

    pub fn add_variable(&mut self, name: EcoString, variable_id: VariableId) {
        self.name_to_variable.insert(name, variable_id);
    }

    pub fn find_variable(&self, name: &EcoString) -> Option<VariableId> {
        if let Some(var_id) = self.name_to_variable.get(name) {
            return Some(*var_id);
        }

        if let ScopeKind::Foreach(var_name, var_id) = &self.kind {
            if name == var_name {
                return Some(*var_id);
            }
        }

        None
    }
}
