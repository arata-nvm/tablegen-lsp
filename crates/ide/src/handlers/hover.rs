use syntax::parser::TextRange;
use syntax::syntax_kind::SyntaxKind;
use syntax::SyntaxNode;

use crate::eval::EvalDatabase;
use crate::file_system::{FilePosition, FileRange};
use crate::symbol_map::{Symbol, SymbolMap};

#[derive(Debug)]
pub struct Hover {
    pub signature: String,
    pub document: Option<String>,
}

pub fn exec(db: &dyn EvalDatabase, pos: FilePosition) -> Option<Hover> {
    let evaluation = db.eval();
    let symbol_map = evaluation.symbol_map();

    let (signature, define_loc) = extract_symbol_signature(symbol_map, pos)?;

    let parse = db.parse(define_loc.file);
    let symbol_doc = extract_doc_comments(parse.syntax_node(), define_loc.range);

    Some(Hover {
        signature,
        document: symbol_doc,
    })
}

fn extract_symbol_signature(
    symbol_map: &SymbolMap,
    pos: FilePosition,
) -> Option<(String, FileRange)> {
    let symbol = symbol_map.find_symbol_at(pos)?;

    let symbol_info = match symbol {
        Symbol::Class(class) => {
            let name = &class.name;
            let template_arg = class
                .iter_template_arg()
                .map(|id| symbol_map.template_arg(id))
                .map(|arg| format!("{} {}", arg.typ, arg.name))
                .collect::<Vec<String>>()
                .join(", ");
            match template_arg.is_empty() {
                false => format!("class {name}<{template_arg}>"),
                true => format!("class {name}"),
            }
        }
        Symbol::TemplateArgument(template_arg) => {
            format!("{} {}", template_arg.typ, template_arg.name)
        }
        Symbol::Field(field) => {
            let parent = symbol_map.symbol(field.parent);
            format!("{} {}::{}", field.typ, parent.name(), field.name)
        }
        Symbol::Def(def) => {
            format!("def {}", def.name)
        }
    };

    let define_loc = symbol.define_loc();

    Some((symbol_info, *define_loc))
}

fn extract_doc_comments(root: SyntaxNode, range: TextRange) -> Option<String> {
    let id_node = root.covering_element(range);
    let identifier_node = match id_node.kind() {
        SyntaxKind::Id => id_node.parent()?,
        SyntaxKind::Identifier => id_node.into_node()?,
        _ => return None,
    };

    // Class or FieldDef or Defset or InnerValue
    let mut parent_node = identifier_node.parent()?;

    if parent_node.kind() == SyntaxKind::InnerValue {
        let value_node = parent_node.parent()?;
        // Def
        parent_node = value_node.parent()?;
    }

    let mut cur_token = parent_node.first_token()?;
    let mut comments = Vec::new();
    loop {
        cur_token = match cur_token.prev_token() {
            Some(t) => t,
            None => break,
        };
        if cur_token.kind() != SyntaxKind::Whitespace || cur_token.text().matches('\n').count() != 1
        {
            break;
        }

        cur_token = match cur_token.prev_token() {
            Some(t) => t,
            None => break,
        };
        if cur_token.kind() != SyntaxKind::LineComment {
            break;
        }

        let comment = cur_token.text();
        if !comment.starts_with("//") {
            break;
        }
        comments.push(comment.trim_start_matches('/').trim_start().to_string());
    }

    let doc = comments.into_iter().rev().collect::<Vec<_>>().join("\n");
    if doc.is_empty() {
        None
    } else {
        Some(doc)
    }
}

#[cfg(test)]
mod tests {
    use crate::tests;

    use super::Hover;

    fn check(s: &str) -> Hover {
        let (db, f) = tests::single_file(s);
        super::exec(&db, f.marker(0)).expect("definition not found")
    }

    #[test]
    fn class() {
        insta::assert_debug_snapshot!(check("class $Foo;"));
    }

    #[test]
    fn template_arg() {
        insta::assert_debug_snapshot!(check("class Foo<int $arg1>;"));
    }

    #[test]
    fn field() {
        insta::assert_debug_snapshot!(check("class Foo { int $field1 = 1; }"));
    }

    #[test]
    fn doc_comment() {
        insta::assert_debug_snapshot!(check(
            r#"
// doc comment
class $Foo;
            "#
        ));
        insta::assert_debug_snapshot!(check(
            r#"
class Foo {
    // doc comment
    int $foo;
}
            "#
        ));
        insta::assert_debug_snapshot!(check(
            r#"
// doc comment
def $foo;
            "#
        ));
        insta::assert_debug_snapshot!(check(
            r#"
// doc comment

class $Foo;
            "#
        ));
    }
}
