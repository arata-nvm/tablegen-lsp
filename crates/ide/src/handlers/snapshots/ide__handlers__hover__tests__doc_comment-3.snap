---
source: crates/ide/src/handlers/hover.rs
expression: "check(r#\"\n// doc comment\ndef $foo;\n            \"#)"
---
Hover {
    signature: "def foo",
    document: Some(
        "doc comment",
    ),
}
